### PR TITLE
Support dynamic charset change on `meta` tags in `HtmlRewriter`.

### PR DESCRIPTION
**Pull request by orium**

### CODE REVIEW

File: src/rewriter/mod.rs [hunk 1]
```diff
@@ -12,13 +12,46 @@ use crate::parser::ParsingAmbiguityError;
...
 use std::error::Error as StdError;
 use std::fmt::{self, Debug};
+use std::ops::Deref;
 use std::rc::Rc;
 use thiserror::Error;
 
 pub use self::settings::*;
 
+#[derive(Clone)]
+pub struct SharedEncoding {
```
Comment (inikulin) [inline src/rewriter/mod.rs:27]: Move to `base` dir

File: src/rewriter/mod.rs [hunk 2]
```diff
@@ -140,6 +173,40 @@ macro_rules! guarded {
     }};
 }
 
+fn ascii_compatible_encoding_from_mimetype(mime: &Mime) -> Option<AsciiCompatibleEncoding> {
```
Comment (inikulin) [inline src/rewriter/mod.rs:176]: make a method on `AsciiCompatibleEncoding`

File: src/rewriter/mod.rs [hunk 3]
```diff
@@ -140,6 +173,40 @@ macro_rules! guarded {
...
+    mime.get_param("charset")
+        .and_then(|cs| Encoding::for_label_no_replacement(cs.as_str().as_bytes()))
+        .and_then(AsciiCompatibleEncoding::new)
+}
+
+fn handler_adjust_charset_on_meta_tag(
```
Comment (inikulin) [inline src/rewriter/mod.rs:182]: Move method after `impl HtmlRewriter {}`

File: src/rewriter/mod.rs [hunk 4]
```diff
@@ -150,12 +217,24 @@ impl<'h, O: OutputSink> HtmlRewriter<'h, O> {
...
-        let encoding = settings.encoding;
+        let encoding = SharedEncoding::new(settings.encoding);
         let mut selectors_ast = selectors_vm::Ast::default();
         let mut dispatcher = ContentHandlersDispatcher::default();
-        let has_selectors = !settings.element_content_handlers.is_empty();
+        let has_selectors =
+            !settings.element_content_handlers.is_empty() || settings.adjust_charset_on_meta_tag;
+
+        let charset_adjust_handler = match settings.adjust_charset_on_meta_tag {
+            true => Some(handler_adjust_charset_on_meta_tag(SharedEncoding::clone(
```
Comment (inikulin) [inline src/rewriter/mod.rs:227]: Just use `if .. else`

File: src/rewriter/mod.rs [hunk 5]
```diff
@@ -558,6 +649,110 @@ mod tests {
...
+        use crate::html_content::{ContentType, TextChunk};
+
+        let enthusiastic_text_handler = || {
+            doc_text!(move |text: &mut TextChunk| {
+                let new_text = text.as_str().replace('!', "!!!");
+                text.replace(&new_text, ContentType::Text);
+                Ok(())
+            })
+        };
+        let html: Vec<u8> = [
```
Comment (inikulin) [inline src/rewriter/mod.rs:663]: Here and in all other places: separate variable declarations, multiline expressions and statements by empty lines

File: src/rewriter/mod.rs [hunk 6]
```diff
@@ -558,6 +649,110 @@ mod tests {
...
+            r#"<meta http-equiv="content-type" content="text/html; charset=windows-1251"><html><head></head><body>I love "#.as_bytes().to_vec(),
+            vec![0xd5, 0xec, 0xb3, 0xcb, 0xdc],
+            r#"!</body></html>"#.as_bytes().to_vec(),
+        ].into_iter().concat();
```
Comment (inikulin) [inline src/rewriter/mod.rs:721]: This needs `cargo fmt`

File: tests/fixtures/token_capturing.rs [hunk 7]
```diff
@@ -2,10 +2,7 @@ use crate::harness::suites::html5lib_tests::{
     get_test_cases, TestCase, TestToken, TestTokenList,
 };
 use crate::harness::{TestFixture, Input};
-use lol_html::{
-    LocalNameHash, TokenCaptureFlags, LocalName, Token, StartTagHandlingResult, TransformController,
-    TransformStream, Namespace, TransformStreamSettings, MemoryLimiter
-};
+use lol_html::{LocalNameHash, TokenCaptureFlags, LocalName, Token, StartTagHandlingResult, TransformController, TransformStream, Namespace, TransformStreamSettings, MemoryLimiter, SharedEncoding};
```
Comment (inikulin) [inline tests/fixtures/token_capturing.rs:5]: This needs cargo fmt. Let's add `cargo fmt -- --check` to CI

**Review by inikulin — CHANGES_REQUESTED** [review]

File: src/rewriter/mod.rs (hunk 4 above)
Comment (orium) [reply src/rewriter/mod.rs:227]: I find matching on booleans to often be more readable than a `if`, specially when both expressions are similar (not the case here).

Happy to change it anyway.

File: src/rewriter/mod.rs (hunk 6 above)
Comment (orium) [reply src/rewriter/mod.rs:721]: This is formatted by `cargo fmt`.

File: tests/fixtures/token_capturing.rs (hunk 7 above)
Comment (orium) [reply tests/fixtures/token_capturing.rs:5]: This is formatted by `cargo fmt`.

> Let's add cargo fmt -- --check to CI

Will open a PR for that.

File: src/rewriter/mod.rs (hunk 1 above)
Comment (orium) [reply src/rewriter/mod.rs:27]: Done.  Does it make sense to also move `AsciiCompatibleEncoding` to `base/encoding.rs`?  We would export it in a different place so it is a breaking change...

Comment (orium) [reply src/rewriter/mod.rs:27]: Actually, `SharedEncoding` is not exported.  It is only used internally.  Do you still want to have it in the `base` module?

File: src/rewriter/mod.rs (hunk 2 above)
Comment (orium) [reply src/rewriter/mod.rs:180]: :heavy_check_mark:

File: src/rewriter/mod.rs (hunk 3 above)
Comment (orium) [reply src/rewriter/mod.rs:186]: :heavy_check_mark:

File: src/rewriter/mod.rs (hunk 4 above)
Comment (orium) [reply src/rewriter/mod.rs:231]: :heavy_check_mark:

File: src/rewriter/mod.rs (hunk 5 above)
Comment (orium) [reply src/rewriter/mod.rs:667]: :heavy_check_mark:

**Review by inikulin — APPROVED** [review]

**Pull request merged**